File: plugin/ftp/www/ftp.js
window.plugin = {
  name: 'FTP',
  loaded: 'ftp_plugin_loaded'
};


function ftp_plugin_loaded() {

	$('#FtpCancelButton').click(function(){
		getFtpData();
		return false;
	});

	$('#FtpSaveButton').click(function(){

		$("body").css("cursor", "progress");

		$.ajax({
			type: "POST",
			url: "/plugin/ftp/ftp.py/set_ftp_data",
			data: _determineDataString(),
			dataType: "json",
			success: function(response) {
				$("body").css("cursor", "default");
				if (response.Result == 'ERROR')
					alert(response.Message);
			},
			error: function (xhr, ajaxOptions, thrownError) {
				$("body").css("cursor", "default");
				alert('Error: ' + xhr.status + ' - ' + thrownError);
			},
		});
		return false;
	});

	$('#FtpTestButton').click(function(){

		$("body").css("cursor", "progress");

		$.ajax({
			type: "POST",
			url: "/plugin/ftp/ftp.py/test_ftp_access",
			data: _determineDataString(),
			dataType: "json",
			success: function(response) {
				$("body").css("cursor", "default");
				if (response.Result == 'ERROR')
					alert(response.Message);
				else
					alert(response.Message + '\n\n' + response.Files + ' files found in directory.');
			},
			error: function (xhr, ajaxOptions, thrownError) {
				$("body").css("cursor", "default");
				alert('Cannot test FTP access: ' + xhr.status + ' - ' + thrownError);
			},
		});
		return false;
	});

	$('#FtpClearDirButton').click(function(){

		if (confirm('Do you really want to delete all files in ' + $('#FtpDirectory').val() + '?')) {
			$("body").css("cursor", "progress");
			$.ajax({
				type: "POST",
				url: "/plugin/ftp/ftp.py/clear_directory",
				data: _determineDataString(),
				dataType: "json",
				success: function(response) {
					$("body").css("cursor", "default");
					alert(response.Message);
				},
				error: function (xhr, ajaxOptions, thrownError) {
					$("body").css("cursor", "default");
					alert('Cannot clear FTP directory: ' + xhr.status + ' - ' + thrownError);
				},
			});
		}
		return false;
	});

	getFtpData();
}


function getFtpData() {
    $.ajax({
        type: "GET",
        url: "/plugin/ftp/ftp.py/get_ftp_data",
        contentType: "application/json; charset=utf-8",
        success: function(response){
			if (response.Result == 'ERROR')
				alert(response.Message);
			else {
	        	$("#FtpActive").prop('checked', response.Active);
	        	$("#FtpHost").val(response.Host);
	        	$("#FtpUser").val(response.User);
	        	$("#FtpPassword").val(response.Password);
	        	$("#FtpDirectory").val(response.Directory);
    		}
        },
		error: function (xhr, ajaxOptions, thrownError) {
			alert('Cannot get FTP data: ' + xhr.status + ' - ' + thrownError);
		},
    });
}


function _determineDataString() {
	return 'active=' + $("#FtpActive").prop('checked') +
	'&host=' + encodeURIComponent($('#FtpHost').val()) +
	'&user=' + encodeURIComponent($('#FtpUser').val()) +
	'&password=' + encodeURIComponent($('#FtpPassword').val()) +
	'&directory=' + encodeURIComponent($('#FtpDirectory').val());
}

File: plugin/ftp/www/ftp.py
# Watchdog - Building supervision with Raspberry Pi
# (C)2017  - Norbert Huffschmid - GNU GPL V3 

import ConfigParser
import ftplib
import json
import re

CONFIG_FILE = '/etc/watchdog.conf'


def get_ftp_data(req):
    req.content_type = 'application/json; charset=UTF8'

    try:
        result = { 'Result': 'OK'}

        config = ConfigParser.SafeConfigParser()
        config.read(CONFIG_FILE)
        result['Active'] = config.getboolean('ftp', 'active')
        result['Host'] = config.get('ftp', 'host')
        result['User'] = config.get('ftp', 'user')
        result['Password'] = config.get('ftp', 'password')
        result['Directory'] = config.get('ftp', 'directory')

        return json.dumps(result)

    except Exception, e:
        return json.dumps({ 'Result': 'ERROR', 'Message': 'Cannot get ftp data: ' + str(e) })


def set_ftp_data(req):
    req.content_type = 'application/json; charset=UTF8'

    try:

        # read POST parameters 
        active = req.form.getfirst('active')
        host = req.form.getfirst('host')
        user = req.form.getfirst('user')
        password = req.form.getfirst('password')
        directory = req.form.getfirst('directory')

        config = ConfigParser.SafeConfigParser()
        config.read(CONFIG_FILE)
        config.set('ftp', 'active', active)
        config.set('ftp', 'host', host)
        config.set('ftp', 'user', user)
        config.set('ftp', 'password', password)
        config.set('ftp', 'directory', directory)
        with open(CONFIG_FILE, 'wb') as configfile:
            config.write(configfile)

        return json.dumps({ 'Result': 'OK' })

    except Exception, e:
        return json.dumps({ 'Result': 'ERROR', 'Message': 'Cannot set ftp data: ' + str(e) })


def test_ftp_access(req):
    req.content_type = 'application/json; charset=UTF8'

    try:

        # read POST parameters 
        host = req.form.getfirst('host')
        user = req.form.getfirst('user')
        password = req.form.getfirst('password')
        directory = req.form.getfirst('directory')

        # perform rudimentary test
        ftp = ftplib.FTP(host)
        ftp.login(user, password)
        ftp.cwd(directory.strip('/'))
        files = [] 
        ftp.dir(files.append)
        ftp.quit()

        return json.dumps({ 'Result': 'OK', 'Message': 'FTP connection successfully tested.' , 'Files': len(files) })

    except Exception, e:
        return json.dumps({ 'Result': 'ERROR', 'Message': 'FTP test failed: ' + str(e) })


def clear_directory(req):
    req.content_type = 'application/json; charset=UTF8'

    try:

        # read POST parameters 
        host = req.form.getfirst('host')
        user = req.form.getfirst('user')
        password = req.form.getfirst('password')
        directory = req.form.getfirst('directory')

        ftp = ftplib.FTP(host)
        ftp.login(user, password)
        ftp.cwd(directory.strip('/'))
        for file in ftp.nlst():
            ftp.delete(file)
        ftp.quit()

        return json.dumps({ 'Result': 'OK', 'Message': 'All files deleted.' })

    except Exception, e:
        return json.dumps({ 'Result': 'ERROR', 'Message': 'Clear command failed: ' + str(e) })
